backend/remote-state/pg/client.go

```diff
@@ -95,70 +63,42 @@ func (c *RemoteClient) Delete() error {
 func (c *RemoteClient) Lock(info *state.LockInfo) (string, error) {
 	var err error
 	var lockID string
-	var txn *sql.Tx
 
 	if info.ID == "" {
 		lockID, err = uuid.GenerateUUID()
 		if err != nil {
 			return "", err
 		}
-		info.Operation = "client"
 		info.ID = lockID
 	}
 
-	// Take exclusive access to the database transaction
-	c.txnMux.Lock()
-	defer c.txnMux.Unlock()
-
-	if c.txn == nil {
-		// Most strict transaction isolation to prevent cross-talk
-		// between incomplete state transactions.
-		txn, err = c.Client.BeginTx(context.Background(), &sql.TxOptions{
-			Isolation: sql.LevelSerializable,
-		})
-		if err != nil {
-			return "", &state.LockError{Info: info, Err: err}
-		}
-		c.txn = txn
-	} else {
-		return "", &state.LockError{Info: info, Err: fmt.Errorf("Client is already in a locking transaction")}
-	}
-
-	// Do not wait before giving up on a contended lock.
-	_, err = c.Client.Exec(`SET LOCAL lock_timeout = 0`)
-	if err != nil {
-		c.rollback(info)
-		return "", err
-	}
-
 	// Try to acquire lock for the existing row.
-	query := `SELECT pg_try_advisory_xact_lock(%s.id) FROM %s.%s WHERE %s.name = $1`
-	row := c.txn.QueryRow(fmt.Sprintf(query, statesTableName, c.SchemaName, statesTableName, statesTableName), c.Name)
-	var didLock []byte
-	err = row.Scan(&didLock)
+	query := `SELECT %s.id, pg_try_advisory_lock(%s.id) FROM %s.%s WHERE %s.name = $1`
+	row := c.Client.QueryRow(fmt.Sprintf(query, statesTableName, statesTableName, c.SchemaName, statesTableName, statesTableName), c.Name)
+	var pgLockId, didLock []byte
+	err = row.Scan(&pgLockId, &didLock)
 	switch {
 	case err == sql.ErrNoRows:
 		// When the row does not yet exist in state, take
 		// the `-1` lock to create the new row.
-		innerRow := c.txn.QueryRow(`SELECT pg_try_advisory_xact_lock(-1)`)
+		innerRow := c.Client.QueryRow(`SELECT pg_try_advisory_lock(-1)`)
```

Comment:
Just to make sure I understand correctly the reason for the 2 lock attempts; this looks for an individual "workspace" entry with a lock, and if it didn't exist it takes a coarser lock on everything to make sure it's safe to create the workspace?

Reply:
Yes, that's right.

Although, what you've pointed out made me realize that there was a gap in this implementation. Once locked for creation [`-1`], after a new workspace's DB row is inserted, another DB session would have been able to lock that now existant row [the row's `id`], before the creation lock was unlocked.

I just pushed another rev that prevents taking a workspace-specific lock if already locked for creation.

Reply:
Thanks for the clarification, this makes sense now, especially with the latest update.